Propose preliminary GC MVP

RossTate opened the pull request:
Please discuss. Start by reading the Overview.

sampsyo commented on proposals/gc/MVP.md:
```diff
@@ -7,314 +7,317 @@ See [overview](Overview.md) for background.
-  - For structures or arrays, it is the RTT provided upon creation, or `anyref` if none.
-  - For `i31ref` references it is the RTT for `i31ref`.
-  - For `null` it is the RTT for `nullref`.
+* In practice, values are expected to be packed when possible. What gets packed will change as engines develop more advanced packing and garbage-collection techniques. At first, the expectation is that small `immutable` schemes with `deep` (or no) equality will be packed. Soon after, schemes with `cases` extension will have those cases encoded as the low bits of a (aligned) pointer. Eventually, s ...
```
A possibly useful expansion here: explain a bit more, from the implementation perspective, how the castability restricts what the RTTI must look like. This current discussion seems to lay things out in a kind of timeline of how "aggressive" an implementation might be with RTTI implementation, but it doesn't quite precisely lay out a map from castability types to possible RTTI implementations.

sampsyo commented on proposals/gc/MVP.md:
```diff
@@ -7,314 +7,317 @@ See [overview](Overview.md) for background.
-  - `struct.set $t i : [(optref $t) ti] -> []`
-    - iff `$t = struct (mut1 t1)^i (var ti) (mut2 t2)*`
-    - and `t = unpacked(ti)`
+* `gcref.get_<numbits> <fieldname>` reads from field `x` of an instance
```
I don't understand the need for this as distinct from `gcref.get`, but maybe it's just a wasm thing.

RossTate replied:
Hmm, now that I think about it, we could just make the output type be the smallest `i32`/`i64` type that fits the bits.

sampsyo commented on proposals/gc/NomOO.md:
```diff
@@ -0,0 +1,515 @@
+                     readable immutable)
+              (field class (gcref $class_class)
+                     readable immutable)
+              (field getClass (func (param (gcref $Object)) (result (gcnref $Class)))
```
I expected the _parameters_ to be `gcnref`s as well, in Java… because every variable is nullable, you'd want it to be allowed to pass a local `gcnref` to one of these methods (although passing null would of course trap).

sampsyo commented on proposals/gc/NomOO.md:
```diff
@@ -0,0 +1,515 @@
+Note that this scheme has no parent.
+This indicates that v-tables are completely separate from objects.
+As such, an engine might manage them very differently, as is common.
+For example, its extensibility attribute is not castable, indicating that the engine does not need to store any casting information in the metadata.
```
Very nice payoff!

sampsyo commented on proposals/gc/NomOO.md:
```diff
@@ -0,0 +1,515 @@
+
+There are two ways in which arrays could be designed.
+Unfortunately, due to either Java's use of covariant arrays or Java's feature of generics, one needs fancy types that are beyond the scope of the MVP to verify the pattern that ensures that a `String[]` contains only `String`s.
+As such, even though code compiled from Java will ensure this invariant, the MVP will have to operate as if any reference array can always contain an arbitrary `Object`.
```
Would the MVP work to make the safer version feasible in a language without generic type erasure and covariant arrays (perhaps C#, for example)? Or is there another limitation I don't quite see?

sampsyo commented on proposals/gc/NomOO.md:
````diff
@@ -0,0 +1,515 @@
+$void_class := ...
+```
+
+In this design, the `$Class` scheme has three cases: primitives (e.g. int, double, boolean&mdash;*not* `Integer`, `Double`, `Boolean`), void, and references (i.e. classes, interfaces, and arrays).
````
I don't entirely see why `$Class` has to have three schemes. Is there a reason the representation of a Class object needs to be different for primitives & void?

sampsyo commented on proposals/gc/TypedFun.md:
```diff
@@ -0,0 +1,346 @@
+
+## Uniform Representation
+
+In order to support parametric polymorphism, the MVP needs to support an efficient uniform representation of the values of a typed functional language. Note, though, that this does not the MVP needs to provide an efficient uniform representation across all languages&mdash;the uniform representation can vary by language because polymorphic functions will only be applied to values from that languag ...
```
```suggestion
In order to support parametric polymorphism, the MVP needs to support an efficient uniform representation of the values of a typed functional language. Note, though, that this does not mean the MVP needs to provide an efficient uniform representation across all languages&mdash;the uniform representation can vary by language because polymorphic functions will only be applied to values from that language. Furthermore, values do not always need to be manipulated in the uniform representation&mdash;many implementations have found it more efficient to manipulate values in specialized representations based on static type information and then explicitly convert to and from the uniform representation when using polymorphic functions.
```

sampsyo commented on proposals/gc/TypedFun.md:
```diff
@@ -0,0 +1,346 @@
+
+## Uniform Representation
+
+In order to support parametric polymorphism, the MVP needs to support an efficient uniform representation of the values of a typed functional language. Note, though, that this does not the MVP needs to provide an efficient uniform representation across all languages&mdash;the uniform representation can vary by language because polymorphic functions will only be applied to values from that languag ...
```
Just for fun, I recently enjoyed reading this proposal for adding unboxed kinds to OCaml. https://github.com/ocaml/RFCs/blob/881b220adc1f358ab15f7743d5cd764222ab7d30/rfcs/unboxed-types.md

RossTate replied:
Cool! Now on my to-read list :)

sampsyo commented on proposals/gc/TypedFun.md:
```diff
@@ -0,0 +1,346 @@
+
+That said, it is still important that all values be expressible in some uniform representation.
+But in designing this uniform representation there is something beyond efficiency that we should consider.
+Just as WebAssembly works hard to develop a design that prevents websites from becoming reliant upon implementation details (like endianness), languages compiling to WebAssembly also do not programs interacting with their compiled modules to become dependent on implementation details.
```
```suggestion
Just as WebAssembly works hard to develop a design that prevents websites from becoming reliant upon implementation details (like endianness), languages compiling to WebAssembly also do not want programs interacting with their compiled modules to become dependent on implementation details.
```

sampsyo commented on proposals/gc/TypedFun.md:
```diff
@@ -0,0 +1,346 @@
+Lastly, each of the closure schemes is `extensible flat`.
+This enables each closure-allocation site in the source code to specify its own child scheme of the appropriate arity.
+The function pointer in the `implementation` field can then cast the given closure argument to its specific associated scheme.
+This both enforces the expectation that the `implementation` is only given associated closure instances, and grants the code referenced by `implementation` access to the fields of the closure that are specfic to that `implementation` (i.e. to the values of the specific local variables that are closed over in the corresponding source code).
```
To be clear, here "this" is the _cast_, right? That is, the need for `implementation` to point to the right closure type is enforced dynamically, not statically. Fancy types could presumably help with that, I assume.

RossTate replied:
Exactly. Should I put another "fancy types" note :)

sampsyo replied:
Probably! :smiley:

sampsyo commented on proposals/gc/TypedFun.md:
```diff
@@ -0,0 +1,346 @@
+Notice that the `tail` field is specifically a `gcref $list`, so lists themselves would be able to internally use this packed representation.
+
+When one `gcref.convert`s such a pointer to `$uniform`, the `nil` could still be packed, but now as some constant that distinguishes from nullary constructors of other algebraic data types, and the singleton `$cons` could be boxed onto the heap.
+The heap representation of `$cons` would need to have a header with a constant that distinguishes it from binary constructors of other algebraic data types.
```
I haven't thought about this too hard, but it seems vaguely disappointing that upcasting to `$uniform` would require additional tagging, especially if there is never a need to `$downcast` again… I'll need to ponder this more.

RossTate replied:
It *might* require additional tagging. Here is an example of how F# uses explicit conversion between packed-pointer representations to get efficiency in the packed case: https://github.com/WebAssembly/gc/issues/53#issuecomment-548902948 I found this to be such a cool example, and so I was surprised that literally no one else followed up on it.

sampsyo replied:
I'll need to look more into this, but for the moment, I couldn't glean from that thread precisely what "the .NET approach" is. I gather that it's not OCaml-style tagging and that it attempts to "recover C-like performance" using "static type information," so it must indeed eschew dynamic information in lots of cases, but I would love to read more about how the type system currently works. (Is it based on your earlier work?)

RossTate replied:
Oh, oops, I forgot that that example builds upon earlier discussion: https://github.com/WebAssembly/gc/issues/53#issuecomment-544620003 https://github.com/WebAssembly/gc/issues/53#issuecomment-545863323 https://github.com/WebAssembly/gc/issues/53#issuecomment-547205663
The gist is that you can use static type information to unbox many values (including algebraic values, as the example shows). But you also need to ensure those values are still representable in places where such static type information isn't available, such as in polymorphic code. In the proposal, having a `$foo` scheme be a child of `$uniform` ensures that "foo" values are representable in the uniform scheme, but making `$foo` an `explicit` child of `$uniform` enables that representation to differ depending on static type information, i.e. when the static type translates to `gcref $foo` (in monomorphic code) versus `gcref $uniform` (in polymorphic code).

sampsyo replied:
Got it; thanks! That's very helpful!

sampsyo commented on proposals/gc/TypedFun.md:
```diff
@@ -0,0 +1,346 @@
+
+> `e1 == e2` tests for physical equality of `e1` and `e2`.
+> On mutable types such as references, arrays, byte sequences, records with mutable fields and objects with mutable instance variables, `e1 == e2` is true if and only if physical modification of `e1` also affects `e2`.
+> On non-mutable types, the behavior of `( == )` is implementation-dependent; however, it is guaranteed that `e1 == e2` implies `compare e1 e2 = 0`.
```
Wow! Did not know this was implementation-dependent.

RossTate replied:
Heheh, oh yes, I learned a lot about how messy OCaml is in developing this in-depth use case for the proposal. Actually, Cornell's own 3110 lecture notes on this are some of the only discussions of the topic. Michael Clarkson and I had fun sharing amusements and frustrations about this :)

RossTate commented on proposals/gc/Overview.md:
```diff
@@ -27,12 +25,16 @@ WARNING: Some contents of this document may have gotten out of sync with the [MV
 * Allocation of byte arrays that are garbage collected
 * Allow heap values from the embedder (e.g. JavaScript objects) that are garbage collected
-* Unboxing of small scalar values
 * Down casts as an escape hatch for the low-level type system
```
In other words, any type system will have limitations, so the design offers down casts as an escape hatch around those limitations (but a dynamically checked one for soundness).

RossTate commented on proposals/gc/Overview.md:
```diff
@@ -27,12 +25,16 @@ WARNING: Some contents of this document may have gotten out of sync with the [MV
-* Unboxing of small scalar values
 * Down casts as an escape hatch for the low-level type system
-* Explicit low-level control over all runtime behaviour (no implicit allocation, no implicit runtime types)
 * Modular (no need for shared type definitions etc.)
```
I believe it is an intentional design goal of WebAssembly, which is why I was irked when I got so much push back after pointing out that the structural design would have to rely heavily on speculative optimizations.

RossTate commented on proposals/gc/Overview.md:
```diff
@@ -77,451 +91,111 @@ GC support should maintain Wasm's efficiency properties as much as possible, nam
-Unless willing to implement runtime code specialisation (like C# / .NET) a type-agnostic compilation scheme is necessary.
+Every scheme specifies a number of attributes, such as the fields describing the data in the scheme, compatibility requirements with other schemes, and the ability to extend the scheme.
+From this information, the engine generates an in-memory representation for the scheme and a packed-pointer representation for the scheme.
+The values of type `gcref $scheme` are packed-pointer representations of instances of `$scheme`.
```
Yeah, I don't have a good point of reference for how to describe this in traditional WebAssembly terms. I'm hoping the wasm team will help us align the terminology and notation with their current norms.

RossTate commented on proposals/gc/Overview.md:
```diff
@@ -77,451 +91,111 @@ GC support should maintain Wasm's efficiency properties as much as possible, nam
-* down casts
+A field might specify a mutation attribute that specifies how the field can change over time.
+Currently the options are `mutable`, `immutable`, and `initializable`.
+The last indicates that the field can only be changed from its default value&mdash;once it has a non-default value, it cannot be changed.
```
Heh, yeah, those sorts of impossible combinations happen all over the place in this proposal. I opted for more readable syntax, but we might actually want to address this issue in the binary format, where impossible combinations mean wasted space.

sabine replied:
While the separate words may seem more readable on first glance, I think it is easier to understand when you cannot emit impossible combinations, as there is less room for confusion. I would suggest cleaning this up in the syntax as well. Also, you don't need to check for impossible combinations when compiling to the binary format if the text syntax doesn't have them in the first place.

RossTate commented:
Thanks, @sampsyo! Your summary of the philosophy seems accurate to me. Glad you like it! @steveblackburn is actually working on making a second pass through the proposal with the specific purpose of evaluating whether it would indeed be compatible with a wide variety of implementation strategies, e.g. precise vs. conservative collectors, bit-flagged vs. header-summarized heap walking, and heavily pointer-packed vs. universally heap-allocated instances.

As for your points: I updated the `indexed` functionality to both incorporate fixed-length indexing (which NomOO already had an example application of) and plan ahead for the possibility of multiple indexing modes. I added to the Approach section of the Overview a discussion of fancy types and looking forward. I'll let you come back to me regarding `implicit` and `explicit`, though some of my replies to your comments might provide some insight into the design feature.

skuzmich commented on proposals/gc/MVP.md:
```diff
@@ -7,314 +7,317 @@ See [overview](Overview.md) for background.
+  - `castattr ::= castable`
+
+* An instantiation attribute indicates how instances of this scheme (as opposed to some child scheme) can be created. A scheme can have at most one construction attribute.
+  - `instantiationattr ::= constructible | extensible (explicit | flat | hierarchical | cases <schemeidx>*)?`
```
Why scheme can't be `constructible` and `extensible` at the same time?

RossTate replied:
I originally allowed that and found it caused a lot of complications and confusions. For example, if I say "instance of `$scheme`", does that mean an instance of exactly `$scheme` or an instance of any child of `$scheme`? It's possible that we'll make both possible at the same time, but for now I felt it would make for any easier discussion with them separate. In the meanwhile, it's easy enough to approximate by the pattern `$constructible := scheme.new (parent implicit $extensible) constructible`.

skuzmich commented on proposals/gc/MVP.md:
```diff
@@ -7,314 +7,317 @@ See [overview](Overview.md) for background.
+    + The `<fieldname>` is used in place of an index so that child schemes can specify fields without knowing the full contents of their parents. The specific name has no run-time significance; it is effectively an abstract name for an unknown offset to be determined by the engine.
+  - `inttype ::= (signed | unsigned) <num-bits>` (where 0 < `<num-bits>` <= 64)
+    + The `signed | unsigned` qualifier informs the engine how the integer should be packed, when appropriate.
+  - `mutationattr ::= mutable | initializable | immutable` 
```
What are the benefits of `initializable` attr over `mutable`?

RossTate replied:
There are three that I had in mind. One is that `initializable` means that subsequent reads of a field that is known to be non-default (e.g. non-null) can be optimized away. Another is that it has a pretty straightforward multi-threading story: just CAS against the default value. The last is that it might enable some useful optimizations for downloading/compiling code on demand, though this is beyond the scope of the current proposal.

skuzmich replied:
> One is that initializable means that subsequent reads of a field that is known to be non-default (e.g. non-null) can be optimized away.
Would you be able to read uninitialized default value? Is so, how would you know if field is initialized to optimize reads? Would you need interpocedural data-flow analysis for that?

RossTate replied:
Suppose the method fields of a v-table were all nullable and `initializable` in order to make room for the possibility that the method implementation was downloaded on demand. A method invocation would then have to check if the value where `null`, in which case proceed with the initialization, before invoking the method. But repeated invocations of the same method could skip the `null` step. Sure, your compiler could optimize this itself. But suppose a general-purpose wasm optimizer comes along and inlines some function call; it won't realize that many of the `null` checks inside can also be optimized away unless it knows the field is only `initializable` and not arbitrarily `mutable`. Similarly, your code generator wouldn't have to apply the optimization at all if it knew a subsequent pass would take care of `initializable` fields for you, making your life easier.

In general, a lot of languages have an initialization phase that WebAssembly needs to support. So I felt this pattern would be common enough that it was worth building it in. It enables a better toolchain and provides engines with some really useful information (such as what the likely hot path will be).

skuzmich commented on proposals/gc/MVP.md:
```diff
@@ -7,314 +7,317 @@ See [overview](Overview.md) for background.
-    - iff `t <: t'`
-  - multiple invocations of this instruction with the same operand yield the same observable RTTs
-  - this is a *constant instruction*
+* `gcnref.eq_distinct` compares two possibly `null` references whose scheme supports equality, treating `null` as unequal to itself
```
What would be the use-cases of treating `null` as unequal to itself?

RossTate replied:
It can be useful in settings where `null` more accurately represents the absence of a value. This is the  semantics of equality in SQL, for example, for precisely this reason.

skuzmich commented on proposals/gc/MVP.md:
```diff
@@ -7,314 +7,317 @@ See [overview](Overview.md) for background.
-    - iff `t' <: t <: anyref`
-    - and `$l : [t']`
-  - branches iff the first operand's runtime type is a sub-RTT of the RTT operand
+* `gcref.convert <schemeidx>` converts an instance to an instance of `$target`
```
Why would we need an explicit up-cast? Wouldn't implicitly using child as a parent be enough?

RossTate replied:
If the parent-child relationship is `implicit`, then this is unnecessary. But if it's `explicit`, then their pointer-packed representation might be different, and so this instruction is converting between these representations. This tends to not be so important for OO languages but important for functional languages and dynamic languages.

skuzmich commented on proposals/gc/Overview.md:
```diff
@@ -77,451 +97,111 @@ GC support should maintain Wasm's efficiency properties as much as possible, nam
+Every scheme specifies a number of attributes, such as the fields describing the data in the scheme, compatibility requirements with other schemes, and the ability to extend the scheme.
+From this information, the engine generates an in-memory representation for the scheme and a packed-pointer representation for the scheme.
+The values of type `gcref $scheme` are packed-pointer representations of instances of `$scheme`.
+If the scheme is `nullable`, then the values of type `gcnref $scheme` are packed-pointer representations of instances of `$scheme` or of `null`.
```
Are there advantages of non-nullable schemes besides not having to reserve encoding for null value?

RossTate replied:
That's the primary advantage I had in mind&mdash;useful for some packed-pointer representations (especially ones that can be packed so much that values are guaranteed to never be allocated on the heap).

RossTate commented:
Thanks, @skuzmich, for the great questions!

Horcrux7 commented:
Great concept. This remove duplicate type checking.

A short question, what is the difference between `scheme.null` and `ref.null` from the reference proposal?

RossTate commented:
> A short question, what is the difference between `scheme.null` and `ref.null` from the reference proposal?

@Horcrux7 Conveniently, I think the answer to your question can be found in the discussion of `gcnull` in the Types section of the design-rationale draft I just pushed.

sampsyo commented on proposals/gc/Rationale.md:
```diff
@@ -0,0 +1,329 @@
+
+First, some meta-level discussion of the role of subtyping in low-level languages:
+* Take a possible instruction design like `struct.set <typeidx> <fieldidx>`. One might think that the `<typeidx>` annotation improves type-checking, but it in fact burdens the type-checker. First, the annotation takes space in the binary while only providing information that can already be deduced from the type of the operand on the stack. Second, the annotation forces the type-checker to check i ...
+* Research has also found that, if one takes things a step further, a well-designed subtyping system can significantly facilitate tooling for low-level languages. If one ensures subtyping is well-founded and has joins, i.e. least upper bounds, then tools can rearrange the control-flow graph and rely on a separate inference tool to automatically recover the critical type annotations at merge point ...
```
Again, we might consider highlighting the top-level outcome in the beginning of the bullet point. Something like this: "A well-designed subtyping system can enable type inference that can eliminate the need to for many explicit types in the instructions. Even if the type inference has to be run offline, to simplify engines, omitting these types can vastly simplify emission and optimization." And also perhaps "In particular, the subtyping relation should be well-founded and have defined least upper bounds for all pairs…"

RossTate replied:
Done!

sampsyo commented on proposals/gc/Rationale.md:
```diff
@@ -0,0 +1,329 @@
+* Take a possible instruction design like `struct.set <typeidx> <fieldidx>`. One might think that the `<typeidx>` annotation improves type-checking, but it in fact burdens the type-checker. First, the annotation takes space in the binary while only providing information that can already be deduced from the type of the operand on the stack. Second, the annotation forces the type-checker to check i ...
+* Research has also found that, if one takes things a step further, a well-designed subtyping system can significantly facilitate tooling for low-level languages. If one ensures subtyping is well-founded and has joins, i.e. least upper bounds, then tools can rearrange the control-flow graph and rely on a separate inference tool to automatically recover the critical type annotations at merge point ...
+* Both of the above imply that one should design subtyping for a low-level language extremely conservatively, only making two types be subtypes if there is a valuable application.
+* One non-application of subtyping in low-level languages is implicit coercion from one type to another. Low-level programs are not generally written by people; they are primarily generated by tools. These tools can easily generate an instruction that explicitly coerces a value from one type to another. This instruction might simply be erased by an engine, or an engine might use static type infor ...
```
Before "This instruction might simply be erased by an engine," maybe it's worth clarifying these points:

- We advocate for explicit conversion instructions when we think _it is likely that some engine will want to use different representations for the two types_.
- Some engines will actually use different representations, but others (perhaps simpler ones) will use the same representation.

Then it's clearer that it's the latter case when the engine can just erase that instruction. But having it is super important for the former case.

I think the point about the unidirectional evolution point at the end is very important. In particular, we could even hammer the point home by pointing out that an engine that _wants to do conversion_ between two representations, if those conversions are implicit, will have a very sad time trying to figure out the optimal points to convert—when emitters could have just told it in the first place with an explicit conversion.

RossTate replied:
Done! Though I compromised by just *emphasizing* the unidirectional text.

RossTate commented on proposals/gc/Rationale.md:
```diff
@@ -0,0 +1,329 @@
+# Subtyping
+
+First, some meta-level discussion of the role of subtyping in low-level languages:
+* Take a possible instruction design like `struct.set <typeidx> <fieldidx>`. One might think that the `<typeidx>` annotation improves type-checking, but it in fact burdens the type-checker. First, the annotation takes space in the binary while only providing information that can already be deduced from the type of the operand on the stack. Second, the annotation forces the type-checker to check i ...
```
Suggestions incorporated! The most relevant paper is my own; it doesn't directly state these things, but it does at least have relevant data.

taralx commented on proposals/gc/MVP.md:
```diff
@@ -7,314 +7,319 @@ See [overview](Overview.md) for background.
-* `rtt t` is a subtype of `anyref`
-  - `rtt t <: anyref`
-  - Note: `rtt t1` is *not* a subtype of `rtt t2`, even if `t1` is a subtype of `t2`; such subtyping would be unsound, since RTTs are used in both co- and contravariant roles (e.g., both when constructing and consuming a reference)
+* `gcnull $scheme1` is a subtype of `gcnull $scheme2` if `$scheme2` is an `implicit` parent or child of `$scheme1` (presuming both `$scheme1` and `$scheme2` are `nullable`).
```
If my type theory is correct, this is identical to having a unit "nullref" type that isn't parameterized by any scheme.

taralx replied:
Ok, not *quite* identical, since under this method `gcnull $scheme1` might be different from `gcnull $scheme2` if the schemes are not related. But is that worth the complexity?

RossTate replied:
One rationale for this is given in the second paragraph of Rationale.md#types. Now that we've talked about roadmapping, another rationale comes from a desire to be forwards compatible with existential types. See MVP.md#approach here for discussion of "fancier type-system features", and Comparison.md#looking-ahead in https://github.com/soil-initiative/gc/pull/2/ for detailed discussion of existential types.

brion commented:
I like the nominal subtyping and the ability to create tagged union ref types I can use to make a universal type representation for a dynamic language, but a couple things concern me:
* whether you get multi-word fat pointers or a tag in the low pointer bits depends entirely on implementation details of the engine, making it hard to make optimization trade-off decisions. This is maybe ok in practice, but can be frustrating when performance cliffs are found accidentally :)
* I don't quite grok the JS interop story. Fat pointers or tagged pointers would presumably not fit in the JS engine's own universal value type, so you'd either need to forbid them from passing the boundary or create a wrapper?
* Further, would field names be exposed as property names, and array index accesses mapped suitably, to JS?

brion commented on proposals/gc/TypedFun.md:
````diff
@@ -0,0 +1,347 @@
+```
+
+Because both of these schemes are entirely `immutable` and permit no notion of equality, the engine can choose to pack common instances of these schemes into the pointer rather than allocate them on the heap.
+Which of the two schemes gets packed will likely depend on the architecture (i.e. 32 vs. 64 bit) and the engine (e.g. pointer flagging vs. NaN boxing).
````
I don't see a way to implement NaN-boxing without additional knowledge being added to the type definition about whether NaN values need to be round-tripped.
